SYCL/ESIMD/radix_sort.cpp

```diff
@@ -0,0 +1,657 @@
+    // WA gather does not take less than 8
+    // simd<unsigned, TUPLE_SZ> rd_off(0,1);  // 0, 1, 2, 3
+    simd<unsigned, 8> rd_off(0, 1);
+    rd_off += (global_offset - TUPLE_SZ);
+    simd<unsigned, 8> temp;
+    temp = gather<unsigned, 8>(buf, rd_off * sizeof(unsigned));
+    prev = temp.select<TUPLE_SZ, 1>(0);
+  }
+
+  simd<unsigned, 32 * TUPLE_SZ> S;
+  unsigned n_iter = n_entries / 32;
+  for (unsigned i = 0; i < n_iter; i++) {
+
+    S = gather_rgba<GATHER_SCATTER_MASK, unsigned int, 32>(buf, element_offset);
+
+    auto cnt_table = S.bit_cast_view<unsigned int, TUPLE_SZ, 32>();
+    cnt_table.column(0) += prev;
+#pragma unroll
+    for (unsigned j = 0; j < TUPLE_SZ; j++) {
+      // step 1
+      cnt_table.select<1, 1, 16, 2>(j, 1) +=
+          cnt_table.select<1, 1, 16, 2>(j, 0);
+      // step 2
+      cnt_table.select<1, 1, 8, 4>(j, 2) += cnt_table.select<1, 1, 8, 4>(j, 1);
+      cnt_table.select<1, 1, 8, 4>(j, 3) += cnt_table.select<1, 1, 8, 4>(j, 1);
+      // step 3
+      cnt_table.select<1, 1, 4, 1>(j, 4) +=
+          cnt_table.replicate_vs_w_hs<1, 0, 4, 0>(j, 3);
+      cnt_table.select<1, 1, 4, 1>(j, 12) +=
+          cnt_table.replicate_vs_w_hs<1, 0, 4, 0>(j, 11);
+      cnt_table.select<1, 1, 4, 1>(j, 20) +=
+          cnt_table.replicate_vs_w_hs<1, 0, 4, 0>(j, 19);
+      cnt_table.select<1, 1, 4, 1>(j, 28) +=
+          cnt_table.replicate_vs_w_hs<1, 0, 4, 0>(j, 27);
+      // step 4
+      cnt_table.select<1, 1, 8, 1>(j, 8) +=
+          cnt_table.replicate_vs_w_hs<1, 0, 8, 0>(j, 7);
+      cnt_table.select<1, 1, 8, 1>(j, 24) +=
+          cnt_table.replicate_vs_w_hs<1, 0, 8, 0>(j, 23);
+      // step 5
+      cnt_table.select<1, 1, 16, 1>(j, 16) +=
+          cnt_table.replicate_vs_w_hs<1, 0, 16, 0>(j, 15);
+    }
+
+    // during reduction phase, we've already computed prefix sum and saved in
+    // the last entry. Here we avoid double counting the last entry
+    if (i == n_iter - 1)
+      cnt_table.column(31) -= cnt_table.column(30);
+
+    scatter_rgba<unsigned int, 32, GATHER_SCATTER_MASK>(buf, element_offset, S);
+
+    element_offset += stride_elems * TUPLE_SZ * sizeof(unsigned) * 32;
+    prev = cnt_table.column(31);
+  }
+}
+
+void hierarchical_prefix(queue &q, unsigned *buf, unsigned elem_stride,
+                         unsigned thrd_stride, unsigned n_entries,
+                         unsigned entry_per_th) {
+  if (n_entries <= REMAINING_ENTRIES) {
+    // std::cout << "... n_entries: " << n_entries << " elem_stide: " <<
+    // elem_stride << " thread_stride: " << thrd_stride << " entry per thread: "
+    // << entry_per_th << std::endl; one single thread
+    q.submit([&](handler &cgh) {
+      cgh.parallel_for<class Accum_final>(
+          range<2>{1, 1} * range<2>{1, 1}, [=](item<2> it) SYCL_ESIMD_KERNEL {
+            cmk_acum_final(buf, it.get_id(0), elem_stride, n_entries);
+          });
+    });
+    return;
+  }
+
+  // std::cout << "*** n_entries: " << n_entries << " elem_stide: " <<
```

Comment:
Nit: maybe cleanup here and line 359

Reply:
The commented lines were added for debugging purpose - removed.